src/charm.py

```
@@ -3,115 +3,267 @@
+
+        self.on.define_event("reconcile_runners", ReconcileRunnersEvent)
+        self.on.define_event("update_runner_bin", UpdateRunnerBinEvent)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.reconcile_runners, self._on_reconcile_runners)
+        self.framework.observe(self.on.update_runner_bin, self._on_update_runner_bin)
+        self.framework.observe(self.on.stop, self._on_stop)
+
         self.framework.observe(
             self.on.check_runners_action, self._on_check_runners_action
         )
-        self.framework.observe(self.on.install, self._on_install)
-        self.framework.observe(self.on.stop, self._on_stop)
-        self.framework.observe(self.on.check_runners, self._on_check_runners)
-        self._stored.set_default(path="")
-        self._runner = Runner(self.config["path"], self.config["token"])
-        self._cron_tab = CronTab
+        self.framework.observe(
+            self.on.reconcile_runners_action, self._on_reconcile_runners_action
+        )
+        self.framework.observe(
+            self.on.flush_runners_action, self._on_flush_runners_action
+        )
+
+    def _get_runner_manager(self, token=None, path=None):
+        """Get a RunnerManager instance, or None if missing config."""
+        if token is None:
+            token = self.config["token"]
+        if path is None:
+            path = self.config["path"]
+        if not (token and path):
+            return None
+        return RunnerManager(path, token, self.app.name, self.config["virt-type"])
 
     def _on_install(self, event):
         self.unit.status = MaintenanceStatus("Installing packages")
-        self._runner.install()
+        RunnerManager.install_deps()
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
```

As much as I like `Option`s from other languages, since you're already using exceptions below, seems like you might want to embrace that and have `self._get_runner_manager()` throw an exception and catch it instead of checking for `None` here.